src/content/popup/KanjiEntryTopPart.tsx

```diff
@@ -0,0 +1,291 @@
+import { KanjiResult } from '@birchill/jpdict-idb';
+
+import { useLayoutEffect, useRef, useState } from 'preact/hooks';
+import { PopupOptions } from './popup';
+import { ReactNode } from 'preact/compat';
+import { useLocale } from '../../common/i18n';
+import { renderFrequency, renderPencil, renderPerson } from './icons';
+
+type Props = {
+  entry: KanjiResult;
+  index: number;
+  options: PopupOptions;
+};
+
+export function KanjiEntryTopPart({ entry, index, options }: Props) {
+  const [lastPointerType, setLastPointerType] = useState<string>('mouse');
+
+  return (
+    <div class="top-part">
+      {/* -- The kanji itself */}
+      <div
+        class="kanji"
+        lang="ja"
+        onPointerUp={({ pointerType }) => {
+          setLastPointerType(pointerType);
+        }}
+        onClick={() => {
+          // TODO: fix this
+          // if (popupHasSelectedText(table)) {
+          //   return;
+          // }
+
+          const trigger = lastPointerType === 'mouse' ? 'mouse' : 'touch';
+          options.onStartCopy?.(index, trigger);
+        }}
+      >
+        {entry.c}
+      </div>
+
+      {/* -- Top-right part */}
+      <div>
+        <Readings entry={entry} />
+
+        {/* -- -- Meta */}
+        {entry.misc.meta && <Meta meta={entry.misc.meta} />}
+
+        {/* -- -- Meanings */}
+        <div className="meanings" lang={entry.m_lang}>
+          {entry.m.join(', ')}
+        </div>
+
+        {/* -- -- Misc info */}
+        <MiscRow entry={entry} />
+
+        {/* -- -- Kanji components */}
+        {(typeof options.showKanjiComponents === 'undefined' ||
+          options.showKanjiComponents) && <KanjiComponents entry={entry} />}
+      </div>
+    </div>
+  );
+}
+
```

Comment:
Not sure what your preference is on helper components being declared above/below the main component. It doesn't particularly matter to me either way, and actually I tend to declare them above, and put the main component at the bottom, but I did the opposite here since that more closely mirrored what you had in the old manual rendering code in `kanji.ts`.

Reply:
Thanks. Yeah, these days I lean towards doing the main thing at the top and adding the helpers below (i.e. leaning on JS's function hoisting).

I think of this as being "newspaper style" where you describe the main thing first then add details below. I find this is easier to read since you start reading at the top, get an overview, and then can continue reading for as long as you're interested.